File: app/src/main/java/ca/rmen/android/poetassistant/widget/CABEditText.java
package ca.rmen.android.poetassistant.widget;

import android.content.Context;
import android.support.v7.widget.AppCompatEditText;
import android.util.AttributeSet;
import android.view.KeyEvent;

/**
 * https://code.google.com/p/android/issues/detail?id=23381
 */
public class CABEditText extends AppCompatEditText implements HackFor23381 {

    public interface ImeListener {
        /**
         * The soft keyboard was just closed.
         */
        void onImeClosed();
    }

    private boolean shouldWindowFocusWait;
    private ImeListener mImeListener;

    public CABEditText(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    public CABEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public CABEditText(Context context) {
        super(context);
    }

    @Override
    public void setWindowFocusWait(boolean shouldWindowFocusWait) {
        this.shouldWindowFocusWait = shouldWindowFocusWait;
    }

    @Override
    public void onWindowFocusChanged(boolean hasWindowFocus) {
        if(!shouldWindowFocusWait) {
            super.onWindowFocusChanged(hasWindowFocus);
        }
    }

    @Override
    public boolean onKeyPreIme(int keyCode, KeyEvent event) {
        if (mImeListener != null
                && event.getKeyCode() == KeyEvent.KEYCODE_BACK
                && event.getAction() == KeyEvent.ACTION_UP) {
            mImeListener.onImeClosed();
        }
        return super.onKeyPreIme(keyCode, event);
    }

    public void setImeListener(ImeListener listener) {
        mImeListener = listener;
    }

}

File: app/src/main/kotlin/ca/rmen/android/poetassistant/widget/CABEditText.kt
package ca.rmen.android.poetassistant.widget

import android.content.Context
import android.support.v7.widget.AppCompatEditText
import android.util.AttributeSet
import android.view.KeyEvent

/**
 * https://code.google.com/p/android/issues/detail?id=23381
 */
class CABEditText : AppCompatEditText, HackFor23381 {

    interface ImeListener {
        /**
         * The soft keyboard was just closed.
         */
        fun onImeClosed()
    }

    constructor(context : Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle)

    constructor(context : Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context) : super(context)

    private var shouldWindowFocusWait: Boolean = false
    var imeListener: ImeListener? = null

    override fun setWindowFocusWait(shouldWindowFocusWait: Boolean) {
        this.shouldWindowFocusWait = shouldWindowFocusWait
    }

    override fun onWindowFocusChanged(hasWindowFocus: Boolean) {
        if (shouldWindowFocusWait) {
            super.onWindowFocusChanged(hasWindowFocus)
        }
    }

    override fun onKeyPreIme(keyCode: Int, event: KeyEvent): Boolean {
        if (imeListener != null
                && event.keyCode == KeyEvent.KEYCODE_BACK
                && event.action == KeyEvent.ACTION_UP) {
            imeListener?.onImeClosed()
        }

        return super.onKeyPreIme(keyCode, event)
    }

}
